### DISCUSSION TITLE
Model for frequent configuration dynamic updates

### DISCUSSION THREAD

**Discussion by vitorpmoreira** [dotnet/yarp#2566] [Q&A] [answered]
```
.NET 8
YARP 2.1.0
```

Hello!
From what I know, dynamic in-memory updates to the YARP configuration requires the user to pass in a snapshot of the Clusters and Routes, and do not support operations like Add, Update or Remove for single entries.
Now, in a use-case where are very frequent updates from different threads, this would require a user to implement thread-safe or locking mechanisms in order to update the model (List), in memory, and then request the YARP update, ensuring consistency and atomicity by blocking other updates to Clusters/Routes that have nothing to do with one-another.

In this case, I have YARP receiving connections from different peers and I want to save each peer's cluster for future outbound requests. I was thinking that, when a peer connects, I would build the Cluster and it's Routes and update Yarp with it, but doesn't seem to be possible.

Am I thinking this in the wrong way and the modelling should be different for these types of use-cases? I already read #2182, and I am using the an `InMemoryProvider` as my single dynamic source.

Thank you and best regards,
Vítor

**Comment by Tratcher** [comment, marked as answer]
Your understanding is sound. InMemoryProvider is a good place to start. You can fork and extend that with Add and Remove methods that update the internal state and signal to YARP that a change is available. Yes this state needs to be managed in a thread safe way, and it needs to provide YARP with a read-only copy each time. Note that you don't have to wait for the YARP update to finish, if another change comes through before the current one is finished it will still be picked up on the next pass. You can cut down on the churn by making those adds and removes batch operations (AddList, RemoveList).
